integration/fuse/src/main/java/alluxio/fuse/file/FuseFileInOrOutStream.java

```diff
@@ -0,0 +1,183 @@
+    }
+    // Left for next operation to decide read-only or write-only mode
+    // read-only: open(READ_WRITE) existing file - read()
+    // write-only: open(READ_WRITE) existing file - truncate(0) - write()
+    return new FuseFileInOrOutStream(fileSystem, authPolicy,
+        Optional.empty(), Optional.empty(), currentStatus, uri, mode);
+  }
+
+  private FuseFileInOrOutStream(FileSystem fileSystem, AuthPolicy authPolicy,
+      Optional<FuseFileInStream> inStream, Optional<FuseFileOutStream> outStream,
+      Optional<URIStatus> status, AlluxioURI uri, long mode) {
+    Preconditions.checkNotNull(fileSystem);
+    Preconditions.checkNotNull(authPolicy);
+    Preconditions.checkNotNull(status);
+    Preconditions.checkNotNull(uri);
+    Preconditions.checkArgument(!(inStream.isPresent() && outStream.isPresent()),
+        "Cannot create both input and output stream");
+    mFileSystem = fileSystem;
+    mAuthPolicy = authPolicy;
+    mOriginalStatus = status;
+    mUri = uri;
+    mMode = mode;
+    mInStream = inStream;
+    mOutStream = outStream;
+  }
+
+  @Override
+  public synchronized int read(ByteBuffer buf, long size, long offset) {
+    if (mOutStream.isPresent()) {
+      throw new UnsupportedOperationException(
+          "Alluxio does not support reading while writing/truncating");
+    }
+    if (!mInStream.isPresent()) {
+      mInStream = Optional.of(FuseFileInStream.create(mFileSystem, mUri,
+          OpenFlags.O_RDONLY.intValue(), mOriginalStatus));
+    }
+    return mInStream.get().read(buf, size, offset);
+  }
+
+  @Override
+  public synchronized void write(ByteBuffer buf, long size, long offset) {
+    if (mInStream.isPresent()) {
+      throw new UnsupportedOperationException(
+          "Alluxio does not support reading while writing/truncating");
+    }
+    if (!mOutStream.isPresent()) {
+      mOutStream = Optional.of(FuseFileOutStream.create(mFileSystem, mAuthPolicy, mUri,
+          OpenFlags.O_WRONLY.intValue(), mMode, mOriginalStatus));
+    }
+    mOutStream.get().write(buf, size, offset);
+  }
+
+  @Override
+  public synchronized long getFileLength() {
+    if (mOutStream.isPresent()) {
+      return mOutStream.get().getFileLength();
+    }
+    if (mInStream.isPresent()) {
+      return mInStream.get().getFileLength();
+    }
+    if (mOriginalStatus.isPresent()) {
+      return mOriginalStatus.get().getLength();
+    }
+    return 0L;
+  }
+
+  @Override
+  public synchronized void flush() {
+    if (mInStream.isPresent()) {
+      mInStream.get().flush();
+      return;
+    }
+    mOutStream.ifPresent(FuseFileOutStream::flush);
+  }
+
+  @Override
+  public synchronized void truncate(long size) {
+    if (mInStream.isPresent()) {
+      throw new UnsupportedOperationException(
+          "Alluxio does not support reading while writing/truncating");
+    }
+    if (mOutStream.isPresent()) {
+      mOutStream.get().truncate(size);
```

Comment:
So as long as we are not done writing, truncate is doable? How often do we see this pattern?

Reply:
Truncate can be called at anytime
For 
`[FuseFileInOrOutStream.java](https://github.com/Alluxio/alluxio/pull/15552/files/398ec62c73b91b996ffeeefc0ce28880e3502913#diff-6e84567f9fd3c47c68c8e9e15a18b242a250bd9aeb027ba8288ebaacd3c7e3b8)`
we let underlying `FuseOutStream` to decide whether it can be truncated or not